### ISSUE TITLE
OTA PIC firmware upgrade

### ISSUE THREAD

**Issue by benjaminvdb** [rvdbreemen/OTGW-firmware#148]
I have an OTGW (PIC16F1847) with a WeMos D1 module from Nodo-Shop. The PIC has firmware 6.2 and the WeMos D1 firmware 0.9.5 (the latest).

Earlier I flashed the PIC firmware using a USB cable, but this only seemed to work if the WeMos was removed from the OTGW board. Now that the OTGW is in a housing, I must disassemble it before I can flash the PIC using a USB cable. It would therefore be great if I could upgrade the PIC firmware OTA instead.

If I go to the "PIC firmware" page of the webUI, I can get the latest version (6.3) by clicking the download button. When I download this file from FSExplorer and compare the md5 checksum with the latest OTGW firmware, I see they are identical. However, the "Device information" page of the webUI still indicates 6.2; this can also be seen via telnet during boot.

How can I flash the latest PIC firmware OTA?

**Comment by Barracuda09** [comment]
Yes, this is something I would like to know as well. I am unable to flash the new PIC firmware 6.3 as well.
I used otmonitor to try to flash the PIC, but it fails to put the PIC in reset mode and says it needs a manually reset, but both do not work.

Were could be the problem? or is it a user error?

**Comment by benjaminvdb** [comment]
@Barracuda09 If you're trying to perform an OTA flash through otmonitor failure is expected as indicated by the documentation:

> Warning: Never flash your OTGW PIC firmware through wifi using OTmonitor application, you can brick your OTGW PIC. Instead use the buildin PIC firmware upgrade feature (based on code by Schelte Bron)

I'm trying to perform an OTA flash through this repositories firmware using the webUI, not otmonitor.

**Comment by Barracuda09** [comment]
Yes, but otmonitor uses OTWG-firmware as well, to perform the PIC firmware flash. But could be a problem bricking if network interruptions occur.

But web interface did not work for me either, that is why I tied the otmonitor path.

**Comment by Barracuda09** [comment]
I think, it may have to do with the WeMos D1 running out of memory trying to allocate mem here:
https://github.com/rvdbreemen/OTGW-firmware/blob/627e3b918562c506e4e54b5dcedb0e0463aa55e1/OTGWSerial.cpp#L460-L468

**Comment by hvxl** [comment]
Upgrading of PIC firmware 6.X fails because the PIC16F1847 uses a slightly different bootloader algorithm than the PIC16F88 due to differences in the hardware. This change has not yet been implemented in the OTGW-firmware for the ESP8266.

Upgrading with OTmonitor fails because the OTGW-firmware collects serial data for a second or until a newline. Bootloader packages end with ETX, not LF. And the bootloader waits up to 1 second for a firmware upgrade to start. So, the OTGW-firmware delays the initial package just long enough for the bootloader to time out.

**Comment by tomm1ed** [comment]
> Upgrading of PIC firmware 6.X fails because the PIC16F1847 uses a slightly different bootloader algorithm than the PIC16F88 due to differences in the hardware. This change has not yet been implemented in the OTGW-firmware for the ESP8266.
> 
> Upgrading with OTmonitor fails because the OTGW-firmware collects serial data for a second or until a newline. Bootloader packages end with ETX, not LF. And the bootloader waits up to 1 second for a firmware upgrade to start. So, the OTGW-firmware delays the initial package just long enough for the bootloader to time out.

So in conclusion OTA flashing is nog supported on PIC16F1847? My OTGW is completely 'build in' so requiring to get it out to flash the PIC is a royal PITA. 
@rvdbreemen do you have any plans on implementing this in the OTGW-Firmware?

Also is there any documentation how to flash through the web interface (I was so far unable to find that ;) )

**Comment by Barracuda09** [comment]
@tomm1ed 

OTA for PIC is possible but not correctly implemented in this firmware. How ever this firmware works https://github.com/hvxl/otgwmcu, but lacks the features that are in this firmware.

**Comment by tomm1ed** [comment]
> https://github.com/hvxl/otgwmcu

Weird I see @rvdbreemen as contributor on that firmware. Can we expect an update to OTGW-Firmware to allow upgrades?

(and also; where is the documentation how an upgrade would go ;) )

**Comment by rvdbreemen** [comment]
> Weird I see @rvdbreemen as contributor on that firmware. Can we expect an update to OTGW-Firmware to allow upgrades?

Yes, that's the plan. Actually @hvxl has given me the code months ago. Life happened. And there is a way around that my firmware does not support the 6.x firmware yet. 

Planning to fix this though. I can't stay behind, need to push several small fixes to the firmware. Plus there is something wrong with the flashing of the current binary.

**Comment by benjaminvdb** [comment]
@rvdbreemen If you open an issue for the issues that you're aware of then perhaps someone else can take a look at it if you like.

**Comment by rvdbreemen** [comment]
@benjaminvdb I would love to have some collaborators. Are you by any change a programmer that can help me out here?

The dev tree with the upgrade code is here: https://github.com/rvdbreemen/OTGW-firmware/tree/dev-prep-6x-upgrades

**Comment by tomm1ed** [comment]
> > Weird I see @rvdbreemen as contributor on that firmware. Can we expect an update to OTGW-Firmware to allow upgrades?
> 
> Yes, that's the plan. Actually @hvxl has given me the code months ago. Life happened. And there is a way around that my firmware does not support the 6.x firmware yet.
> 
> Planning to fix this though. I can't stay behind, need to push several small fixes to the firmware. Plus there is something wrong with the flashing of the current binary.

Ah wonderfull! Looking forward to any updates you can supply. I know how life can be ;)

**Comment by 0crap** [comment]
Subscribed.

My recently obtained OTGW also has a PIC16F1847 running firmware 6.2
I'll wait trying v6.3 until this issue is fixed.

**Comment by rvdbreemen** [comment]
@benjaminvdb the dev-prep-6x-upgrades was the branch I was working on. Good to hear you are a programmer too. I do develop everything in the open and accept Pull Requests. This summer I was working on integrating it, you probably looked at that branch. 

I love to see some development on the firmware, so you are more then welcome to check out my 6x upgrade branch. And no, I did not get it to work. Due to personal reasons, I stopped this summer, other stuff had more priority. I am taking my Christmas vacation and hope to spend some time on my project again.

So if you feel you want to help me out, please do.

**Comment by benjaminvdb** [comment]
@rvdbreemen What I meant to say with regards to transparency is perhaps more an issue of organization. In my mind - but correct me if I'm wrong - there is working code capable of upgrading the PIC in Schelte's codebase. However, I don't think I have access to it; I found only firmware binaries on his website. If I have his exact code then I can try to integrate it.

For me, a partial integration is not an ideal starting point. I found it hard to understand the aim of the development branch and had the impression it was trying to do multiple things. I guess that's fine if commits are squashed and traceable to individual issues, but I don't think that's currently the case. I'd suggest to create individual issues for bugs, features, etc. and keep relevant information in there so it's easier for others to contribute. I understand it requires some effort, but it could alleviate the workload in the long run. You've created something wonderful and useful here and I'm sure others will want to contribute to keep it alive and kicking!

In short, could you share the relevant upgrade code from Schelte's codebase? I promise to then take a stab at it!

I wish you a wonderful Christmas 🎅

**Comment by hvxl** [comment]
The library is part of my public otgwmcu project: https://github.com/hvxl/otgwmcu/tree/master/libraries/OTGWSerial

**Comment by benjaminvdb** [comment]
@hvxl That's awesome. Thanks!  :pray:

**Comment by Barracuda09** [comment]
I have looked to this branch https://github.com/rvdbreemen/OTGW-firmware/tree/dev-prep-6x-upgrades  as well.

My problem was that the malloc failed to allocate memory. See comment: see https://github.com/rvdbreemen/OTGW-firmware/issues/148#issuecomment-1339828542

There is a lot of 'debug' strings that takes up a lot of memory. 

So it would maybe wise to clean up this first? or some other method to put it into flash memory.

**Comment by Barracuda09** [comment]
I have to dive into this some more, but what is better was my problem. This _OTWG-firmware_ or _otwgmcu_.
This firmware support more like FSExplorer, MQTT and commands etc. (only not yet PIC firmware update)

I personally only use MQTT the rest is nice to have. So I stopped at this point to see were my time would be wise to invest.

**Comment by 0crap** [comment]
MQTT is probably the most commonly used feature, in combination with Home automation software.
I would vote for the MQTT stuff and the possibility to keep the PIC up to date.
If that works, add missing stuff in small iterations?

But I'm not qualified to answer the question, maybe this approach breaks everything. ;-)

**Comment by Barracuda09** [comment]
The main reason I like to update the PIC firmware OTA was to modify the PIC firmware.

**Comment by rvdbreemen** [comment]
Thanks for all the input on what you would like to see working. I would simply try to fix the OTA upgrading first. The UI was designed for use on a smartphone in my case. I get why you like other UI, it looks a lot like the OTmonitor software  too.

**Comment by rvdbreemen** [comment]
So over the last few weeks I have been trying to finally get back into getting this fixed. First I had to update a lot of library stuff, but that's done now. So I am back to about where I stopped before the summer vacation, trying to make 6.x upgrades work using @hvxl excellent serial class.

In this thread there have been several hints on what might be the issue. The problem is actually that the new OTGWSerial.cpp library claims 8k of memory (malloc that @Barracuda09 pointed out), this is needed for the 6.x firmware. This however is not possible, it fails to allocate memory and thus no upgrade ever happens.

So now I have to see if I can reduce the memory footprint, without loosing functionality. Hmmm, any suggestions?

I see @0crap and @Barracuda09 already figured this out a while ago. Now dropping some functionality might be an option, but where do I start. I have to think about it. I agree that MQTT and Home Assistant integration are the core of my firmware, it always was about that. The REST API is used to communicate with the Web pages, and can be used to control the device.

Like I said, I need to think about this, upgrading to a ESP32 compatible device might be a way to go, with WeMos that is possible ;-)

**Comment by tjfsteele** [comment]
If you can stream the firmware from the filesystem instead of keeping it in
memory..?

Tim

**Comment by benjaminvdb** [comment]
> In this thread there have been several hints on what might be the issue. The problem is actually that the new OTGWSerial.cpp library claims 8k of memory (malloc that @Barracuda09 pointed out), this is needed for the 6.x firmware. This however is not possible, it fails to allocate memory and thus no upgrade ever happens.

@hvxl Is 8k of memory actually required for the current 6.x firmware or is this to prepare the code for larger firmware sizes in the future? If not, what is the amount of memory currently required? I've tried to figure this out from the binary file produced by hex2bin but I was unable to figure it out.

**Comment by hvxl** [comment]
Not the full 8k is required. But the code is not placed in one contiguous area by the linker. The PIC goto and call instructions can only address a 2k block. Therefor the linker must place each code package in its entirety inside a single block. With the latest firmware, the linker has placed code in address ranges 0x000000 - 0x00077b, 0x000800 - 0x000a43, and 0x001800 - 0x001ea5. The self-programming code is in 0x001f00 - 0x001fe9. So allocating only the required amount of memory is going to be a bit complicated.

**Comment by rvdbreemen** [comment]
What if we write out the hex translation to a binary file. And from there we stream it to the PIC over serial. That way the impact on the code is small. 

PS. In the mean time I will go and move strings to flash memory to save space...

**Comment by rvdbreemen** [comment]
Link to file format explainer: www.elproducts.com/understanding-hex-files.html

**Comment by Barracuda09** [comment]
Hi @rvdbreemen 

Yes, I am familiar with this hex file format and programmers. But for some reason I think it could be done with less memory allocation.

But I have for testing here moved a lot of strings (or removed) to flash memory to clear (heap) memory.

**Comment by hvxl** [comment]
There are several reasons for reading the complete file. In addition to what is already mentioned:
* Determine for which PIC the firmware is intended.
* The PICs must be programmed in blocks of 32 words. The hex file does not necessarily provide the code aligned to these blocks, or even in sequential order. Programming a block in parts would erase what was programmed before, leading to corruption of the firmware.
* Figure out the metrics for the progress indication.

So, reading the complete code into memory was the easiest solution. Doing it in two passes is very likely possible. But it will be much more complicated. Especially if a code block is revisited later in the hex file. It may be necessary to disallow that. But it still has to be detected.

**Comment by rvdbreemen** [comment]
> But I have for testing here moved a lot of strings (or removed) to flash memory to clear (heap) memory. 

I did the same thing last night. To no resort. I got 16kB of free memory. 

> But that seem not to work out as well, this could well be a lack of the web interface progress bar

It just failed with me. You can check the debug logs for our of memory error.

**Comment by akula3** [comment]
Hi,
I bricked the P16F1847 by using OTmonitor via OTA. I know, it is my fault :(
A replacement already ordered at nodoshop.
Nevertheless, is there a possibility to recover the bricked PIC?
Thank you for your valuable suggestions!

**Comment by tjfsteele** [comment]
You may be able to recover it using otmonitor via usb

**Comment by Barracuda09** [comment]
Hi @akula3 

Yes, you have a change to use: https://github.com/hvxl/otgwmcu
to "unbrick" your PIC. Just flash this to your NodeMCU and update your PIC from there. I had success with flashing the PIC (It was not bricked).
It seems to work quite well and I believe it will also work when the main part is flashed wrong, at least if the self programming code is not erased

**Comment by rvdbreemen** [comment]
@Barracuda09 how much memory did you win by dumping removing strings and stuff? Just curious how far you got with this.

**Comment by Barracuda09** [comment]
@rvdbreemen, I do not know how many exactly. But at least enough to allocate the memory for OTGWUpgradeData.

I placed most of the text/strings between F() macro and removed(empty) defines such as #define MQTTDebugTln(...)

**Comment by akula3** [comment]
@Barracuda09, thank you for pointing me into the right direction!
Now the PIC is healthy again, eventough I did it in slightly different way you suggested.
I downloaded fw.bin and fs.bin from http://www.otgw.tclcode.com/download.html#otgwmcu and uploaded into D1 mini, then connected to OTmonitor via USB and uploaded there the newest PIC firmware 6.4
Thanks once more!

**Comment by Barracuda09** [comment]
Hi @akula3 credits go out to @tjfsteele as well for the USB suggestion.

**Comment by rvdbreemen** [comment]
The v0.10.0 firmware has been released and now flashes both pic16f1848 (6.x firmware) as pic16f88 (5.x firmware).  I think we can close this issue now ;-)

**Comment by Barracuda09** [comment]
Hi @rvdbreemen 

This seem to work to update. But how can I update the PIC with a self build hex file? or am I missing something?

**Comment by benjaminvdb** [comment]
> The v0.10.0 firmware has been released and now flashes both pic16f1848 (6.x firmware) as pic16f88 (5.x firmware). I think we can close this issue now ;-)

This is incredible work, @rvdbreemen. It works flawlessly! 🙌

**Comment by rvdbreemen** [comment]
@Barracuda09 currently the filebrowser does not go into directories. That means you cannot manually upload your custom hex file. The only way is to create a custom FS using the Arduino plugin yourself that includes your custom hexfile.

**Comment by hvxl** [comment]
Maybe a simpler fix/work-around than making the file browser handle subdirectories, could be to search for hex files in the pic-specific directory **and** the root directory.

@Barracuda09, the latest firmware does allow upgrading the PIC with OTmonitor. So that's another option. Of course it's a bit more risky if your WiFi is not rock solid.

**Comment by Barracuda09** [comment]
Hi @rvdbreemen & @hvxl 

Thanks for you feedback and suggestions. I will try this out.

Thanks for your support!!
